File: python/year_2015/day_19.py
"""
https://adventofcode.com/2015/day/19
"""

import re
from collections import defaultdict
from itertools import chain, starmap
from pathlib import Path
from random import shuffle

from iteration_utilities import empty


def part_1(text):
    def replace(i, token):
        for replacement in replacements.get(token, empty):
            tokens[i] = replacement
            yield "".join(tokens)
        tokens[i] = token

    def tokenize(molecule):
        return re.findall(r"[A-Z][a-z]?", molecule)

    lines, molecule = text.strip().split("\n\n")

    replacements = defaultdict(list)
    for line in lines.split("\n"):
        key, val = line.split(" => ")
        replacements[key].append(val)

    tokens = tokenize(molecule)

    return len(set(chain.from_iterable(starmap(replace, enumerate(tokens)))))


def part_2(text, starting="e"):
    def process(line):
        return line.split(" => ")

    lines, molecule = text.strip().split("\n\n")
    replacements = list(map(process, lines.split("\n")))
    saved, steps = molecule, 0

    while molecule != starting:
        for x, y in replacements:
            if y not in molecule:
                continue

            molecule = molecule.replace(y, x, 1)
            steps += 1
            break
        else:
            shuffle(replacements)
            molecule, steps = saved, 0

    return steps


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_19.txt").read_text()


EXAMPLE_1 = """H => HO
H => OH
O => HH

HOH
"""

EXAMPLE_2 = """e => H
e => O
H => HO
H => OH
O => HH

HOH
"""


def test_part_1_input():
    assert part_1(_read_input()) == 576


def test_part_1_example_1():
    assert part_1(EXAMPLE_1) == 4


def test_part_1_example_2():
    assert part_1(EXAMPLE_2) == 4


def test_part_2_input():
    assert part_2(_read_input()) == 207


def test_part_2_example_1():
    assert part_2(EXAMPLE_1, "O") == 2


def test_part_2_example_2():
    assert part_2(EXAMPLE_2) == 3

File: rust/year_2015/src/day_19.rs
use hashbrown::HashSet;

pub fn part_1(input: &str) -> usize {
    let (molecule, replacements) = parse_input(input);
    replacements
        .flat_map(|(from, to)| {
            molecule.match_indices(from).map(move |(i, _)| {
                let mut molecule = molecule.to_string();
                molecule.replace_range(i..i + from.len(), to);
                molecule
            })
        })
        .collect::<HashSet<_>>()
        .len()
}

pub fn part_2(input: &str) -> usize {
    let (molecule, _) = parse_input(input);
    let n = molecule.matches(char::is_uppercase).count();
    let rn = molecule.matches("Rn").count();
    let ar = molecule.matches("Ar").count();
    let y = molecule.matches("Y").count();
    n - rn - ar - 2 * y - 1
}

fn parse_input(input: &str) -> (&str, impl Iterator<Item = (&str, &str)>) {
    let (replacements, molecule) = input.rsplit_once("\n\n").unwrap();
    let replacements = replacements
        .lines()
        .map(|line| line.split_once(" => ").unwrap());
    (molecule, replacements)
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_19.txt");

    const EXAMPLE_1: &str = "H => HO
H => OH
O => HH

HOH";

    const EXAMPLE_2: &str = "H => HO
H => OH
O => HH

HOHOHO";

    #[test_case(EXAMPLE_1 => 4)]
    #[test_case(EXAMPLE_2 => 7)]
    #[test_case(INPUT => 576)]
    fn part_1(input: &str) -> usize {
        super::part_1(input)
    }

    //     const EXAMPLE_3: &str = "e => H
    // e => O
    // H => HO
    // H => OH
    // O => HH
    //
    // HOH";

    //     const EXAMPLE_4: &str = "e => H
    // e => O
    // H => HO
    // H => OH
    // O => HH
    //
    // HOHOHO";

    // #[test_case(EXAMPLE_3 => 3)]
    // #[test_case(EXAMPLE_4 => 6)]
    #[test_case(INPUT => 207)]
    fn part_2(input: &str) -> usize {
        super::part_2(input)
    }
}
